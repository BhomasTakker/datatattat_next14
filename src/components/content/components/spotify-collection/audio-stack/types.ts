import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
