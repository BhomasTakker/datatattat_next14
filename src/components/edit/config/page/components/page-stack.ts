import { EditInputs } from "@/components/edit/inputs/inputs";
import { InputListProps } from "@/types/edit/inputs/inputs";

export const PAGE_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Page Stack",

	inputs: [],
};
