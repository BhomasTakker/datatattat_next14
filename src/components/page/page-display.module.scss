@use "../../scss/abstracts/units/" as *;

.page {
	display: flex;
	flex-direction: column;
}
