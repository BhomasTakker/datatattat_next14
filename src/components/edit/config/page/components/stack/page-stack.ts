import { EditInputs } from "@/components/edit/inputs/inputs";
import { PageStackCollectionVariants } from "@/types/components/page/stack";
import { InputListProps } from "@/types/edit/inputs/inputs";
import { ROW_STACK } from "./row-stack-variant";
import { VERTICAL_STACK } from "./vertical-stack-variant";

const pageStackMap = new Map<string, InputListProps>([
	[PageStackCollectionVariants.Vertical, VERTICAL_STACK],
	[PageStackCollectionVariants.RowStack, ROW_STACK],
]);

export const PAGE_STACK_CONFIG: InputListProps = {
	id: "stack",
	type: EditInputs.inputList,
	label: "Page Stack",

	inputs: [
		{
			id: "variant",
			type: EditInputs.objectSelect,
			label: "Stack Variant",
			options: [...Object.values(PageStackCollectionVariants)],
			optionMap: pageStackMap,
			defaultValue: PageStackCollectionVariants.Vertical,
		},
	],
};
