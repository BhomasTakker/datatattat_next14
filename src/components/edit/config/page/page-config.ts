import { GenericInput } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../inputs/inputs";
import { COMPONENT_CONFIG } from "../components/component-config";
import { PAGE_GRID_CONFIG } from "./components/page-grid";
import { PAGE_STACK_CONFIG } from "./components/stack/page-stack";

enum pageContainers {
	STACK = "Stack",
	GRID = "Grid",
}

type pageContainersProps = typeof PAGE_STACK_CONFIG | typeof PAGE_GRID_CONFIG;

const pageContainersMap = new Map<string, pageContainersProps>([
	[pageContainers.STACK, PAGE_STACK_CONFIG],
	[pageContainers.GRID, PAGE_GRID_CONFIG],
]);

export const PAGE_CONFIG: GenericInput = {
	// id used will be content
	id: "page",
	type: EditInputs.inputList,
	label: "Page",

	inputs: [
		{
			// add button here to save as page template - no we would want to save meta and page properties too no?
			// add button here to demo page
			id: "title",
			type: EditInputs.title,
			title: "Add Page interaction buttons here",
		},
		{
			id: "containerType",
			type: EditInputs.objectSelect,
			label: "Page Container",
			defaultValue: pageContainers.STACK,
			required: true,
			options: [pageContainers.STACK],
			optionMap: pageContainersMap,
			// save to/create on props object
			optionId: "props",
		},
		COMPONENT_CONFIG,
	],
};
