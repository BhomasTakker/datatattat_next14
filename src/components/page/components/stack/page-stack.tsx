import { PageComponents, PageContainer, PageContent } from "@/types/page";
import styles from "./page-stack.module.scss";
import { ComponentDisplay } from "@/components/content/component-display";

// We should probably have 'style' in the Props type
type PageStackProps = {
	direction: string;
};

export type PageStackContent = {
	container: PageContainer;
	props: PageStackProps;
	components: PageComponents;
};

export const PageStack = ({ content }: { content: PageContent }) => {
	const { props, components } = content as PageStackContent;

	const { direction: style = "column" } = props || {};

	////////////////////////////////////////////////////////
	// Look at articleCollection for what we shuld do here
	////////////////////////////////////////////////////////
	const renderComponents = () => {
		return components.map((component, index) => {
			return (
				<li key={index}>
					<ComponentDisplay key={index} component={component} />
				</li>
			);
		});
	};

	const className = `${styles.root} ${styles[`root-${style}`]}`;

	// Semantically this should be a list of components
	return <ul className={className}>{renderComponents()}</ul>;
};
