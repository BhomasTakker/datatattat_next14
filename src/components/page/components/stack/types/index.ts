import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
