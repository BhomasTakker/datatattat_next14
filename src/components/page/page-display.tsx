import { IPage } from "@/types/page";
import { PageProfile } from "./profile/page-profile";
import { PageComponentFactory } from "./components/page-component-factory";
import styles from "./page-display.module.scss";
import { getCurrentRoute } from "@/utils/route";

export const PageDisplay = async ({ page }: { page: IPage }) => {
	const { content, meta, profile } = page || {};

	// not here
	const currentRoute = await getCurrentRoute();
	const isUserRoute = currentRoute?.includes("/users/") || false;
	const isClient = isUserRoute;

	return (
		<main className={styles.page}>
			{/* REPLACE with meta function */}
			{/* Page head for meta data, seo, etc */}
			{/* <PageHead headData={meta} /> */}
			{/* Page profile data for title, description, createdBy, etc */}
			<PageProfile profile={profile} />
			<PageComponentFactory content={content} isClient={isClient} />
		</main>
	);
};
