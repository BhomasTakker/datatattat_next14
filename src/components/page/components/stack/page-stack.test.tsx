import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { PageStack, PageStackContent } from "./page-stack";
import { PageComponent } from "@/types/page";

jest.mock("../../../content/component-display", () => {
	return {
		__esModule: true,
		ComponentDisplay: ({ component }: { component: PageComponent }) => {
			const { componentType, componentProps } = component;
			return (
				<>
					{`Component type:- ${componentType}`}
					<section>{`Component props:- ${JSON.stringify(
						componentProps
					)}`}</section>
				</>
			);
		},
	};
});

const MOCK: PageStackContent = {
	container: {
		containerType: "Stack",
		initData: {},
	},
	props: {
		direction: "column",
	},
	components: [],
};

const MOCK_COMPONENT = {
	componentType: "ListItem",
	componentProps: {
		variant: "default",
		componentTitle: "Sample Title",
		componentTitleLink: "/sample-link",
		showComponentTitle: true,
	},
	_with: {
		type: "sampleType",
		query: {
			key: "sampleKey",
			value: "sampleValue",
			provider: "sampleProvider",
			params: {},
			conversions: [
				{
					response: "sampleResponse",
					sub: "sampleSub",
					conversionId: "sampleConversionId",
					additionalProp: "sampleAdditionalProp",
				},
			],
			queryId: "sampleQueryId",
		},
	},
};

describe("Page Stack Test Suite", () => {
	it("should render an unordered list", () => {
		render(<PageStack content={MOCK} />);
		const component = screen.getByRole("list");
		expect(component).toBeInTheDocument();
	});

	it("should render an unordered list with expected class", () => {
		render(<PageStack content={MOCK} />);
		const component = screen.getByRole("list");
		expect(component).toHaveClass("root", "root-column");
	});

	it("should render a list item", () => {
		// @ts-expect-error whatever this is
		MOCK.components = [MOCK_COMPONENT];
		render(<PageStack content={MOCK} />);
		const component = screen.getByRole("listitem");
		expect(component).toBeInTheDocument();
	});

	it("should render expected number of list items", () => {
		// @ts-expect-error whatever this is
		MOCK.components = [MOCK_COMPONENT, MOCK_COMPONENT, MOCK_COMPONENT];
		render(<PageStack content={MOCK} />);
		const components = screen.getAllByRole("listitem");
		expect(components).toHaveLength(3);
	});

	it("renders PageStack unchanged", () => {
		// @ts-expect-error whatever this is
		MOCK.components = [MOCK_COMPONENT];
		const { container } = render(<PageStack content={MOCK} />);
		expect(container).toMatchSnapshot();
	});
});
