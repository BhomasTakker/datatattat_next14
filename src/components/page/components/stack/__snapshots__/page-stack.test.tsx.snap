// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Page Stack Test Suite renders PageStack unchanged 1`] = `
<div>
  <ul
    class="root"
    data-testid="page-component"
  >
    <li
      class="
		item
	}"
      data-testid="content-component"
      style="--component-height: 100%;"
    >
      Component type:- ListItem
      <section>
        Component props:- {"variant":"default","componentTitle":"Sample Title","componentTitleLink":"/sample-link","showComponentTitle":true}
      </section>
    </li>
  </ul>
</div>
`;
